File: backend-nestjs/src/modules/zones/application/queries/get-zone-by-id.handler.ts
import { IQueryHandler, QueryHandler } from '@nestjs/cqrs';
import { Inject, NotFoundException } from '@nestjs/common';
import type { Cache } from 'cache-manager';
import { CACHE_MANAGER } from '@nestjs/cache-manager';
import { GetZoneByIdQuery } from './get-zone-by-id.query';
import { ZoneRepository } from '../../infrastructure/zone.repository';
import { Zone } from '../../domain/zone';

@QueryHandler(GetZoneByIdQuery)
export class GetZoneByIdHandler implements IQueryHandler<
  GetZoneByIdQuery,
  Zone
> {
  constructor(
    private readonly repository: ZoneRepository,
    @Inject(CACHE_MANAGER) private readonly cache: Cache,
  ) {}

  async execute(query: GetZoneByIdQuery): Promise<Zone> {
    const key = `zone:${query.id}`;
    const cached = await this.cache.get<Zone>(key);
    if (cached) return cached;
    const zone = await this.repository.findById(query.id);
    if (!zone) throw new NotFoundException(`Zone ${query.id} not found`);
    await this.cache.set(key, zone);
    return zone;
  }
}

File: backend-springboot/src/main/java/com/greenalgeria/zone/application/query/GetZoneByIdHandler.java
package com.greenalgeria.zone.application.query;

import com.greenalgeria.shared.cqrs.QueryHandler;
import com.greenalgeria.zone.application.*;
import com.greenalgeria.zone.domain.ZoneRepository;
import java.util.Optional;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
public class GetZoneByIdHandler implements QueryHandler<GetZoneByIdQuery, Optional<ZoneResponse>> {

    private final ZoneRepository zoneRepository;

    public GetZoneByIdHandler(ZoneRepository zoneRepository) {
        this.zoneRepository = zoneRepository;
    }

    @Transactional(readOnly = true)
    public Optional<ZoneResponse> handle(GetZoneByIdQuery query) {
        return zoneRepository.findById(query.id()).map(ZoneResponse::from);
    }

    @Override
    public Class<GetZoneByIdQuery> supportedQuery() {
        return GetZoneByIdQuery.class;
    }
}
